### ISSUE TITLE
"Compile and Run" problem with Java 9

### ISSUE THREAD

**Issue by imagingbook** [imagej/imagej1#52]
Testing ImageJ1 (1.51u) with Java 9 (jre-9.0.4), I noticed that ``Compile and Run`` is not working for plugins defined in a package other than the default package. In this case a ``StackOverflowError`` is shown:
````
Exception in thread "My_First_Plugin" java.lang.StackOverflowError
	at java.base/java.io.WinNTFileSystem.normalizePrefix(Unknown Source)
	at java.base/java.io.WinNTFileSystem.normalize(Unknown Source)
	at java.base/java.io.WinNTFileSystem.normalize(Unknown Source)
	at java.base/java.io.File.<init>(Unknown Source)
	at java.base/jdk.internal.loader.URLClassPath$FileLoader.<init>(Unknown Source)
	at java.base/jdk.internal.loader.URLClassPath$3.run(Unknown Source)
	at java.base/jdk.internal.loader.URLClassPath$3.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(Unknown Source)
	at java.base/jdk.internal.loader.URLClassPath.getLoader(Unknown Source)
	at java.base/jdk.internal.loader.URLClassPath.getResource(Unknown Source)
	at java.base/java.net.URLClassLoader$1.run(Unknown Source)
	at java.base/java.net.URLClassLoader$1.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at ij.plugin.PlugInExecuter.runCompiledPlugin(Compiler.java:335)
	at ij.plugin.PlugInExecuter.runCompiledPlugin(Compiler.java:357)
	at ij.plugin.PlugInExecuter.runCompiledPlugin(Compiler.java:357)
	at ij.plugin.PlugInExecuter.runCompiledPlugin(Compiler.java:357)
	at ij.plugin.PlugInExecuter.runCompiledPlugin(Compiler.java:357)
	at ij.plugin.PlugInExecuter.runCompiledPlugin(Compiler.java:357)
        ...
````
This is the same on Windows7 and MacOs. Compile and run works fine as long as the plugin class is in the default package. The problem appears only when the plugin is defined inside a package (subfolder). In this case, ImageJ's class loader appears to delegate class loading recursively, until the Java stack is full and the runtime crashes.

**Comment by imagejan** [comment]
This was [cross-posted to the ImageJ mailing list](http://imagej.1557.x6.nabble.com/ImageJ1-Compile-and-Run-problem-with-Java-9-tp5020200.html). Please disclose when cross-posting, so others can follow the discussion as well.

**Comment by rasband** [comment]
With Java 9 (jre 9.0.4), Compile and Run does not work for me with plugins in the default package. I created a Java 9 package on OS X by duplicating the ImageJ folder and replacing the Java 8 jre with the Java 9 jre. When I use Compile and Run I get this message

![screenshot](https://user-images.githubusercontent.com/1633637/37014719-87d0a202-20d0-11e8-8a73-6a6621104a0e.png)

After clicking "OK" I get the following exception. It looks like Compile and Run is trying to use the compiler included with the JDK installed on my system (1.8.0_45).

```
> An exception has occurred in the compiler (1.8.0_45). Please file a bug at the Java Developer Connection (http://java.sun.com/webapps/bugreport)  after checking the Bug Parade for duplicates. Include your program and the following diagnostic in your report.  Thank you.
> java.lang.NullPointerException
> 	at com.sun.tools.javac.file.Locations.getPathEntries(Locations.java:149)
> 	at com.sun.tools.javac.file.Locations.getPathEntries(Locations.java:134)
> 	at com.sun.tools.javac.file.Locations.access$000(Locations.java:71)
> 	at com.sun.tools.javac.file.Locations$BootClassPathLocationHandler.computePath(Locations.java:617)
> 	at com.sun.tools.javac.file.Locations$BootClassPathLocationHandler.lazy(Locations.java:643)
> 	at com.sun.tools.javac.file.Locations$BootClassPathLocationHandler.isDefault(Locations.java:536)
> 	at com.sun.tools.javac.file.Locations.isDefaultBootClassPath(Locations.java:109)
> 	at com.sun.tools.javac.file.JavacFileManager.isDefaultBootClassPath(JavacFileManager.java:162)
> 	at com.sun.tools.javac.main.JavaCompiler.<init>(JavaCompiler.java:402)
> 	at com.sun.tools.javac.main.JavaCompiler.instance(JavaCompiler.java:88)
> 	at com.sun.tools.javac.main.Main.compile(Main.java:485)
> 	at com.sun.tools.javac.main.Main.compile(Main.java:381)
> 	at com.sun.tools.javac.main.Main.compile(Main.java:370)
> 	at com.sun.tools.javac.main.Main.compile(Main.java:361)
> 	at com.sun.tools.javac.Main.compile(Main.java:74)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
> 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
> 	at CompilerTool$LegacyCompilerTool.compile(Compiler.java:371)
> 	at Compiler.compile(Compiler.java:95)
> 	at Compiler.compileAndRun(Compiler.java:59)
> 	at Compiler.run(Compiler.java:36)
> 	at ij.IJ.runUserPlugIn(IJ.java:221)
> 	at ij.IJ.runPlugIn(IJ.java:185)
> 	at ij.IJ.runPlugIn(IJ.java:174)
> 	at ij.plugin.Compiler.compileAndRun(Compiler.java:65)
> 	at ij.plugin.Compiler.run(Compiler.java:37)
> 	at ij.IJ.runPlugIn(IJ.java:191)
> 	at ij.IJ.runPlugIn(IJ.java:174)
> 	at ij.plugin.frame.Editor.compileAndRun(Editor.java:379)
> 	at ij.plugin.frame.Editor.actionPerformed(Editor.java:693)
> 	at java.desktop/java.awt.MenuItem.processActionEvent(Unknown Source)
> 	at java.desktop/java.awt.MenuItem.processEvent(Unknown Source)
> 	at java.desktop/java.awt.MenuComponent.dispatchEventImpl(Unknown Source)
> 	at java.desktop/java.awt.MenuComponent.dispatchEvent(Unknown Source)
> 	at java.desktop/java.awt.MenuComponent.dispatchEventImpl(Unknown Source)
> 	at java.desktop/java.awt.MenuComponent.dispatchEvent(Unknown Source)
> 	at java.desktop/java.awt.EventQueue.dispatchEventImpl(Unknown Source)
> 	at java.desktop/java.awt.EventQueue.access$500(Unknown Source)
> 	at java.desktop/java.awt.EventQueue$3.run(Unknown Source)
> 	at java.desktop/java.awt.EventQueue$3.run(Unknown Source)
> 	at java.base/java.security.AccessController.doPrivileged(Native Method)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
> 	at java.desktop/java.awt.EventQueue$4.run(Unknown Source)
> 	at java.desktop/java.awt.EventQueue$4.run(Unknown Source)
> 	at java.base/java.security.AccessController.doPrivileged(Native Method)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
> 	at java.desktop/java.awt.EventQueue.dispatchEvent(Unknown Source)
> 	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
> 	at java.desktop/java.awt.EventDispatchThread.run(Unknown Source)
> 
```

**Comment by imagingbook** [comment]
I get the very same error with Java9/Win7 when trying to actually **compile** a plugin (.java) file. Interestingly, it also refers to the 1.8.0_45 compiler, although there is no such runtime installed on my system.
Did you try ``Compile and Run`` on the (pre-compiled) .class file?

**Comment by rasband** [comment]
The bug  on Java 9 with "Compile and Run" and pre-compiled .class files is fixed in the latest ImageJ daily build (1.51v27). The format of the NoClassDefFoundError error message changed with Java 9. The problem with compiling plugins remains. The reference to 1.8.0_45 is probably becauset Compiler.jar was compiled using that version of Java.

**Comment by imagingbook** [comment]
Thanks for fixing this so promptly, Wayne! I hope the remaining compiler problem can resoved too.
